// storybook-addon-react-docgen/dist/Components/PropVal.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactAddonsCreateFragment = _interopRequireDefault(require("react-addons-create-fragment"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

var styles = {
  "default": {
    color: '#1EA7FD',
    fontWeight: 700,
    wordBreak: 'break-word'
  },
  empty: {
    color: '#444'
  }
};

function indent(breakIntoNewLines, level, isBlock) {
  return breakIntoNewLines && _react["default"].createElement("span", null, _react["default"].createElement("br", null), "".concat(new Array(level).join('  '), "  "), !isBlock && '  ');
}

function PreviewArray(_ref) {
  var val = _ref.val,
      level = _ref.level,
      maxPropArrayLength = _ref.maxPropArrayLength,
      maxPropStringLength = _ref.maxPropStringLength,
      maxPropsIntoLine = _ref.maxPropsIntoLine;
  var items = {};
  var breakIntoNewLines = val.length > maxPropsIntoLine;
  val.slice(0, maxPropArrayLength).forEach(function (item, i) {
    items["n".concat(i)] = _react["default"].createElement("span", null, indent(breakIntoNewLines, level), _react["default"].createElement(PropVal, {
      val: item,
      level: level + 1,
      maxPropStringLength: maxPropStringLength,
      maxPropsIntoLine: maxPropsIntoLine
    }));
    items["c".concat(i)] = ',';
  });

  if (val.length > maxPropArrayLength) {
    items.last = _react["default"].createElement("span", null, indent(breakIntoNewLines, level), '…');
  } else {
    delete items["c".concat(val.length - 1)];
  }

  return _react["default"].createElement("span", {
    style: styles["default"]
  }, "[", (0, _reactAddonsCreateFragment["default"])(items), indent(breakIntoNewLines, level, true), "]");
}

PreviewArray.propTypes = {
  val: _propTypes["default"].any,
  // eslint-disable-line react/require-default-props
  maxPropArrayLength: _propTypes["default"].number.isRequired,
  maxPropStringLength: _propTypes["default"].number.isRequired,
  maxPropsIntoLine: _propTypes["default"].number.isRequired,
  level: _propTypes["default"].number.isRequired
};

function PreviewObject(_ref2) {
  var val = _ref2.val,
      level = _ref2.level,
      maxPropObjectKeys = _ref2.maxPropObjectKeys,
      maxPropStringLength = _ref2.maxPropStringLength,
      maxPropsIntoLine = _ref2.maxPropsIntoLine;
  var names = Object.keys(val);
  var items = {};
  var breakIntoNewLines = names.length > maxPropsIntoLine;
  names.slice(0, maxPropObjectKeys).forEach(function (name, i) {
    items["k".concat(i)] = _react["default"].createElement("span", null, indent(breakIntoNewLines, level), _react["default"].createElement("span", {
      style: styles["default"]
    }, name));
    items["c".concat(i)] = ': ';
    items["v".concat(i)] = _react["default"].createElement(PropVal, {
      val: val[name],
      level: level + 1,
      maxPropStringLength: maxPropStringLength,
      maxPropsIntoLine: maxPropsIntoLine
    });
    items["m".concat(i)] = ',';
  });

  if (names.length > maxPropObjectKeys) {
    items.rest = _react["default"].createElement("span", null, indent(breakIntoNewLines, level), '…');
  } else {
    delete items["m".concat(names.length - 1)];
  }

  return _react["default"].createElement("span", {
    style: styles["default"]
  }, '{', (0, _reactAddonsCreateFragment["default"])(items), indent(breakIntoNewLines, level, true), '}');
}

PreviewObject.propTypes = {
  // eslint-disable-next-line react/require-default-props
  val: _propTypes["default"].any,
  maxPropObjectKeys: _propTypes["default"].number.isRequired,
  maxPropStringLength: _propTypes["default"].number.isRequired,
  maxPropsIntoLine: _propTypes["default"].number.isRequired,
  level: _propTypes["default"].number.isRequired
};

function PropVal(props) {
  var level = props.level,
      maxPropObjectKeys = props.maxPropObjectKeys,
      maxPropArrayLength = props.maxPropArrayLength,
      maxPropStringLength = props.maxPropStringLength,
      maxPropsIntoLine = props.maxPropsIntoLine;
  var val = props.val;
  var content = null;

  if (typeof val === 'number') {
    content = _react["default"].createElement("span", {
      style: styles["default"]
    }, val);
  } else if (typeof val === 'string') {
    if (val.length > maxPropStringLength) {
      val = "".concat(val.slice(0, maxPropStringLength), "\u2026");
    }

    if (level > 1) {
      val = "'".concat(val, "'");
    }

    content = _react["default"].createElement("span", {
      style: styles["default"]
    }, val);
  } else if (typeof val === 'boolean') {
    content = _react["default"].createElement("span", {
      style: styles["default"]
    }, "".concat(val));
  } else if (Array.isArray(val)) {
    content = _react["default"].createElement(PreviewArray, {
      val: val,
      level: level,
      maxPropArrayLength: maxPropArrayLength,
      maxPropStringLength: maxPropStringLength,
      maxPropsIntoLine: maxPropsIntoLine
    });
  } else if (typeof val === 'function') {
    content = _react["default"].createElement("span", {
      style: styles["default"]
    }, val.name || 'anonymous');
  } else if (!val) {
    content = _react["default"].createElement("span", {
      style: styles["default"]
    }, "".concat(val));
  } else if (_typeof(val) !== 'object') {
    content = _react["default"].createElement("span", null, "\u2026");
  } else if (_react["default"].isValidElement(val)) {
    content = _react["default"].createElement("span", {
      style: styles["default"]
    }, "<".concat(val.type.displayName || val.type.name || val.type, " />"));
  } else {
    content = _react["default"].createElement(PreviewObject, {
      val: val,
      level: level,
      maxPropObjectKeys: maxPropObjectKeys,
      maxPropStringLength: maxPropStringLength,
      maxPropsIntoLine: maxPropsIntoLine
    });
  }

  return content;
}

PropVal.defaultProps = {
  val: null,
  maxPropObjectKeys: 3,
  maxPropArrayLength: 3,
  maxPropStringLength: 50,
  maxPropsIntoLine: 3,
  level: 1
};
PropVal.propTypes = {
  val: _propTypes["default"].any,
  maxPropObjectKeys: _propTypes["default"].number,
  maxPropArrayLength: _propTypes["default"].number,
  maxPropStringLength: _propTypes["default"].number,
  maxPropsIntoLine: _propTypes["default"].number,
  level: _propTypes["default"].number
};
var _default = PropVal;
exports["default"] = _default;
module.exports = exports.default;
