// storybook-addon-react-docgen/dist/Components/PropTable.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = PropTable;
exports.multiLineText = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _PropVal = _interopRequireDefault(require("./PropVal"));

var _PrettyPropType = _interopRequireDefault(require("./types/PrettyPropType"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var cell = {
  paddingRight: 20,
  paddingTop: 15,
  paddingBottom: 15,
  verticalAlign: 'top',
  border: 'none'
};
var styles = {
  table: {
    width: '100%',
    margin: '2rem 0',
    borderCollapse: 'collapse'
  },
  header: {
    paddingRight: 20,
    paddingBottom: 10,
    textAlign: 'left',
    fontWeight: 'bold',
    fontSize: 14,
    whiteSpace: 'nowrap',
    border: 'none',
    borderBottom: '1px solid #ccc'
  },
  property: _objectSpread({}, cell, {
    fontWeight: 500,
    color: '#FF4400'
  }),
  propType: _objectSpread({}, cell, {
    fontWeight: 500,
    maxWidth: '150px',
    overflow: 'auto',
    color: '#66BF3C'
  }),
  required: _objectSpread({}, cell),
  defaultValue: _objectSpread({}, cell),
  description: _objectSpread({}, cell)
};

var multiLineText = function multiLineText(input) {
  if (!input) {
    return input;
  }

  var text = String(input);
  var arrayOfText = text.split(/\r?\n|\r/g);
  var isSingleLine = arrayOfText.length < 2;

  if (isSingleLine) {
    return text;
  }

  return arrayOfText.map(function (lineOfText, i) {
    return (// eslint-disable-next-line react/no-array-index-key
      _react["default"].createElement("span", {
        key: "".concat(lineOfText, ".").concat(i)
      }, i > 0 && _react["default"].createElement("br", null), " ", lineOfText)
    );
  });
};

exports.multiLineText = multiLineText;

var determineIncludedPropTypes = function determineIncludedPropTypes(propDefinitions, excludedPropTypes) {
  if (excludedPropTypes.length === 0) {
    return propDefinitions;
  }

  return propDefinitions.filter(function (propDefinition) {
    return !excludedPropTypes.includes(propDefinition.property);
  });
};

function PropTable(props) {
  var type = props.type,
      maxPropObjectKeys = props.maxPropObjectKeys,
      maxPropArrayLength = props.maxPropArrayLength,
      maxPropStringLength = props.maxPropStringLength,
      propDefinitions = props.propDefinitions,
      excludedPropTypes = props.excludedPropTypes;

  if (!type) {
    return null;
  }

  var includedPropDefinitions = determineIncludedPropTypes(propDefinitions, excludedPropTypes);

  if (includedPropDefinitions.length === 0) {
    return _react["default"].createElement("small", null, "No propTypes defined!");
  }

  var propValProps = {
    maxPropObjectKeys: maxPropObjectKeys,
    maxPropArrayLength: maxPropArrayLength,
    maxPropStringLength: maxPropStringLength
  };
  return _react["default"].createElement("table", {
    style: styles.table
  }, _react["default"].createElement("thead", null, _react["default"].createElement("tr", null, _react["default"].createElement("th", {
    style: styles.header
  }, "property"), _react["default"].createElement("th", {
    style: styles.header
  }, "propType"), _react["default"].createElement("th", {
    style: styles.header
  }, "required"), _react["default"].createElement("th", {
    style: styles.header
  }, "default"), _react["default"].createElement("th", {
    style: styles.header
  }, "description"))), _react["default"].createElement("tbody", null, includedPropDefinitions.map(function (row) {
    return _react["default"].createElement("tr", {
      key: row.property
    }, _react["default"].createElement("td", {
      style: styles.property
    }, row.property), _react["default"].createElement("td", {
      style: styles.propType
    }, _react["default"].createElement(_PrettyPropType["default"], {
      propType: row.propType
    })), _react["default"].createElement("td", {
      style: styles.required
    }, row.required ? 'yes' : '-'), _react["default"].createElement("td", {
      style: styles.defaultValue
    }, row.defaultValue === undefined ? '-' : _react["default"].createElement(_PropVal["default"], _extends({
      val: row.defaultValue
    }, propValProps))), _react["default"].createElement("td", {
      style: styles.description
    }, multiLineText(row.description)));
  })));
}

PropTable.displayName = 'PropTable';
PropTable.defaultProps = {
  type: null,
  propDefinitions: [],
  excludedPropTypes: []
};
PropTable.propTypes = {
  type: _propTypes["default"].func,
  maxPropObjectKeys: _propTypes["default"].number.isRequired,
  maxPropArrayLength: _propTypes["default"].number.isRequired,
  maxPropStringLength: _propTypes["default"].number.isRequired,
  excludedPropTypes: _propTypes["default"].arrayOf(_propTypes["default"].string),
  propDefinitions: _propTypes["default"].arrayOf(_propTypes["default"].shape({
    property: _propTypes["default"].string.isRequired,
    propType: _propTypes["default"].oneOfType([_propTypes["default"].object, _propTypes["default"].string]),
    required: _propTypes["default"].bool,
    description: _propTypes["default"].string,
    defaultValue: _propTypes["default"].any
  }))
};
