// storybook-addon-react-docgen/dist/Components/Story.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.getProps = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactLifecyclesCompat = require("react-lifecycles-compat");

var _propTypes = _interopRequireDefault(require("prop-types"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; if (obj != null) { var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

var getName = function getName(type) {
  return type.displayName || type.name;
};

var getDescription = function getDescription(type) {
  return type.__docgenInfo && type.__docgenInfo.description;
};

var getProps = function getProps(propTables, propTablesExclude, children) {
  var types = new Map();

  if (propTables === null) {
    return null;
  }

  if (propTables) {
    propTables.forEach(function (type) {
      types.set(type, true);
    });
  }

  var propTableCompare = function propTableCompare(element, Component) {
    // https://github.com/gaearon/react-hot-loader#checking-element-types
    // eslint-disable-next-line no-undef
    if (typeof reactHotLoaderGlobal !== 'undefined') {
      // eslint-disable-next-line no-undef
      if (reactHotLoaderGlobal.areComponentsEqual(element.type, Component)) {
        return true;
      }
    }

    if (element.type === Component) {
      return true;
    }

    if (element.type && _typeof(element.type) && element.type.name === Component) {
      return true;
    }

    return false;
  }; // Depth-first traverse and collect types


  var extract = function extract(innerChildren) {
    if (!innerChildren) {
      return;
    }

    if (Array.isArray(innerChildren)) {
      innerChildren.forEach(extract);
      return;
    }

    if (innerChildren.props && innerChildren.props.children) {
      extract(innerChildren.props.children);
    }

    if (typeof innerChildren === 'string' || typeof innerChildren.type === 'string' || Array.isArray(propTablesExclude) && // Also ignore excluded types
    propTablesExclude.some(function (Comp) {
      return propTableCompare(innerChildren, Comp);
    })) {
      return;
    }

    if (innerChildren.type && !types.has(innerChildren.type)) {
      types.set(innerChildren.type, true);
    }
  }; // Extract components from children


  extract(children);

  var array = _toConsumableArray(types.keys());

  array.sort(function (a, b) {
    return getName(a) > getName(b);
  });
  return array;
};

exports.getProps = getProps;
var stylesheetBase = {
  infoBody: {
    fontWeight: 300,
    lineHeight: 1.45,
    fontSize: '15px',
    padding: '20px 40px 40px',
    borderRadius: '2px',
    marginTop: '20px',
    marginBottom: '20px'
  },
  h1: {
    margin: '20px 0 0 0',
    padding: '0 0 5px 0',
    fontSize: '25px',
    borderBottom: '1px solid #EEE'
  },
  propTableHead: {
    margin: '20px 0 0 0',
    fontSize: 20
  },
  description: {}
};

var Story =
/*#__PURE__*/
function (_Component) {
  _inherits(Story, _Component);

  function Story() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, Story);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(Story)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_this), "state", {
      stylesheet: _this.props.styles(stylesheetBase)
    });

    return _this;
  }

  _createClass(Story, [{
    key: "_renderInline",
    value: function _renderInline() {
      var stylesheet = this.state.stylesheet;
      return _react["default"].createElement("div", null, _react["default"].createElement("div", {
        style: stylesheet.infoPage
      }, _react["default"].createElement("div", {
        style: stylesheet.infoBody
      }, this._getPropTables())));
    }
  }, {
    key: "_getPropTables",
    value: function _getPropTables() {
      var _this2 = this;

      var _this$props = this.props,
          components = _this$props.components,
          maxPropObjectKeys = _this$props.maxPropObjectKeys,
          maxPropArrayLength = _this$props.maxPropArrayLength,
          maxPropStringLength = _this$props.maxPropStringLength,
          excludedPropTypes = _this$props.excludedPropTypes;
      var propTables = this.props.propTables;
      var stylesheet = this.state.stylesheet;

      if (propTables === null) {
        return null;
      }

      propTables = components.map(function (type, i) {
        var description = getDescription(type);
        return (// eslint-disable-next-line react/no-array-index-key
          _react["default"].createElement("div", {
            key: "".concat(getName(type), "_").concat(i)
          }, _react["default"].createElement("h2", {
            style: stylesheet.propTableHead
          }, "\u201C", getName(type), "\u201D Component"), description && _react["default"].createElement("p", {
            style: stylesheet.description
          }, description), _react["default"].createElement(_this2.props.PropTable, {
            type: type,
            maxPropObjectKeys: maxPropObjectKeys,
            maxPropArrayLength: maxPropArrayLength,
            maxPropStringLength: maxPropStringLength,
            excludedPropTypes: excludedPropTypes
          }))
        );
      });

      if (!propTables || propTables.length === 0) {
        return null;
      }

      return _react["default"].createElement("div", null, _react["default"].createElement("h1", {
        style: stylesheet.h1
      }, "Prop Types"), propTables);
    }
  }, {
    key: "render",
    value: function render() {
      return this._renderInline();
    }
  }]);

  return Story;
}(_react.Component);

_defineProperty(Story, "displayName", 'Story');

_defineProperty(Story, "propTypes", {
  propTables: _propTypes["default"].arrayOf(_propTypes["default"].func),
  styles: _propTypes["default"].func.isRequired,
  components: _propTypes["default"].array.isRequired,
  maxPropObjectKeys: _propTypes["default"].number.isRequired,
  maxPropArrayLength: _propTypes["default"].number.isRequired,
  maxPropStringLength: _propTypes["default"].number.isRequired,
  excludedPropTypes: _propTypes["default"].arrayOf(_propTypes["default"].string)
});

_defineProperty(Story, "defaultProps", {
  propTables: null,
  excludedPropTypes: []
});

(0, _reactLifecyclesCompat.polyfill)(Story);
var _default = Story;
exports["default"] = _default;
